File: src/main/java/ting/leetcode/java/linkedlist/RemoveNthNodeFromEndOfList.java
package ting.leetcode.java.linkedlist;

import org.jetbrains.annotations.Nullable;

public class RemoveNthNodeFromEndOfList {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummyHead = new ListNode();
        dummyHead.next = head;
        ListNode targetPreviousNode = findNthNodeFromEnd(dummyHead, n + 1);
        targetPreviousNode.next = targetPreviousNode.next.next;
        return dummyHead.next;
    }

    @Nullable
    private ListNode findNthNodeFromEnd(ListNode head, int n) {
        // 這個演算法的核心精髓就是間距的概念
        // 利用雙指針先拉出跟題目要求一模一樣的間距，然後同時移動兩根指針
        // 直到第一根指針達到尾端，第二根指針跟第一根指針的間距，就是尾部節點到達倒數第 n 個節點的距離。
        ListNode fast = head;
        ListNode slow = head;
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }

        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        return slow;
    }
}

File: src/main/kotlin/ting/leetcode/linkedlist/RemoveNthNodeFromEndOfList.kt
package ting.leetcode.linkedlist

class RemoveNthNodeFromEndOfList {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        var head = head
        if (head?.next == null) return null
        var slowPointer = head
        var fastPointer = head
        var count = 1
        while (fastPointer?.next != null) {
            fastPointer = fastPointer.next
            count++
            if (count > n && fastPointer?.next != null) {
                slowPointer = slowPointer?.next
            }
        }
        if (count == n) {
            head = slowPointer?.next
            return head
        }

        val nextToTargetNode = slowPointer?.next?.next
        slowPointer?.next = nextToTargetNode

        return head
    }


    fun removeNthFromEndNew(head: ListNode?, n: Int): ListNode? {
        val dummyHead = ListNode(0)
        dummyHead.next = head
        val targetPreviousNode = findNthFromEnd(dummyHead, n + 1)
        targetPreviousNode.next = targetPreviousNode.next!!.next
        return dummyHead.next
    }

    fun findNthFromEnd(head: ListNode, n: Int): ListNode {
        var fast: ListNode? = head
        var slow: ListNode? = head

        for (i in 0 until n) {
            fast = fast!!.next
        }

        while (fast != null) {
            fast = fast.next
            slow = slow!!.next
        }

        return slow!!
    }
}
